### PR TITLE
Fix BotoCredentialAdapter

### PR DESCRIPTION
**Pull request by jeffrey856**
Attempts to fix Cactus that is trying to pickle an open Boto connection

### CODE REVIEW

File: src/toil/__init__.py [hunk 1]
```diff
@@ -456,214 +452,223 @@ def str_to_datetime(s):
...
+                        log.debug('Wrote credentials to %s. Renaming to %s.', tmp_path, path)
+                    os.rename(tmp_path, path)
+                    return
+                finally:
+                    if fd is not None:
+                        os.close(fd)
+                        
+
+# Now we have defined the adapter class. Patch the Boto module so it replaces the default Provider when Boto makes Providers.
+provider.Provider = BotoCredentialAdapter
```
Comment (adamnovak) [inline src/toil/__init__.py]: This line should stay inside the monkey patching function.

File: src/toil/__init__.py [hunk 2]
```diff
@@ -456,214 +452,223 @@ def str_to_datetime(s):
...
-    class BotoCredentialAdapter(provider.Provider):
-        """
-        Adapter to allow Boto 2 to use AWS credentials obtained via Boto 3's
-        credential finding logic. This allows for automatic role assumption
-        respecting the Boto 3 config files, even when parts of the app still use
-        Boto 2.
+
+from boto import provider
+
+class BotoCredentialAdapter():
```
Comment (adamnovak) [inline src/toil/__init__.py]: This still needs to inherit from the provider class.

File: src/toil/__init__.py [hunk 3]
```diff
@@ -456,214 +452,223 @@ def str_to_datetime(s):
...
         return datetime.strptime(s, datetime_format)
 
-    class BotoCredentialAdapter(provider.Provider):
-        """
-        Adapter to allow Boto 2 to use AWS credentials obtained via Boto 3's
-        credential finding logic. This allows for automatic role assumption
-        respecting the Boto 3 config files, even when parts of the app still use
-        Boto 2.
+
+from boto import provider
```
Comment (adamnovak) [inline src/toil/__init__.py]: You can't import from boto at module scope if boto isn't installed. You need something like:

```
try:
    from boto import provider
    base_class = provider.Provider
except ImportError:
    base_class = object

class BotoCredentialAdapter(base_class):
```

File: src/toil/__init__.py [hunk 4]
```diff
@@ -456,215 +453,229 @@ def str_to_datetime(s):
...
+    credential finding logic. This allows for automatic role assumption
+    respecting the Boto 3 config files, even when parts of the app still use
+    Boto 2.
+    
+    This class also handles cacheing credentials in multi-process environments
+    to avoid loads of processes swamping the EC2 metadata service.
+    """
+    
+    from botocore.session import Session
+    from botocore.credentials import create_credential_resolver, RefreshableCredentials, JSONFileCache
```
Comment (adamnovak) [inline src/toil/__init__.py]: You might want to move these imports into the `try` above. I don't think they'll work here.

**Comment by jeffrey856** [comment]
new clean PR at #2904

**Pull request closed without merging**